epochStart/metachain/sovereignRewards.go

```diff
@@ -0,0 +1,201 @@
+package metachain
+
+import (
+	"bytes"
+	"math/big"
+	"sort"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/data"
+	"github.com/multiversx/mx-chain-core-go/data/block"
+	"github.com/multiversx/mx-chain-core-go/data/rewardTx"
+	"github.com/multiversx/mx-chain-go/epochStart"
+	"github.com/multiversx/mx-chain-go/process"
+	"github.com/multiversx/mx-chain-go/state"
+)
+
+type fillBaseRewardsPerBlockPerNodeFunc func(baseRewardsPerNode *big.Int)
+
+type sovereignRewards struct {
+	*rewardsCreatorV2
+}
+
+// NewSovereignRewards creates a sovereign rewards computer
+func NewSovereignRewards(rc *rewardsCreatorV2) (*sovereignRewards, error) {
+	if check.IfNil(rc) {
+		return nil, process.ErrNilRewardsCreator
+	}
+
+	sr := &sovereignRewards{
+		rc,
+	}
+
+	sr.fillBaseRewardsPerBlockPerNodeFunc = sr.fillBaseRewardsPerBlockPerNode
+	sr.flagDelegationSystemSCEnabled.SetValue(true)
+	return sr, nil
+}
+
+// CreateRewardsMiniBlocks creates the rewards miniblocks according to economics data and validator info.
+// This method applies the rewards according to the economics version 2 proposal, which takes into consideration
+// stake top-up values per node. For each created reward tx, it will add directly to the reward txs pool
+func (rc *sovereignRewards) CreateRewardsMiniBlocks(
+	metaBlock data.MetaHeaderHandler,
+	validatorsInfo state.ShardValidatorsInfoMapHandler,
```

Comment:
Nil check missing for `validatorsInfo`

Reply:
There is no nil check for it anywhere in the code, not even in original code, but we can add it here just to make sure